File: Recursion_and_Backtracking_pepcoding_level_1/1.Recursion_Intro/4.factorial.js
const factorial = (n) => {
    if (n === 0 || n === 1) return 1;

    const faith = factorial(n - 1); //faith
    const ans = n * faith; // faith + expection

    return ans;
};

console.log(factorial(5));

// Expectation
// factorial(3) = 3 * 2 * 1;

// Faith
// factorial(2) = 2 * 1;

// F + E
// factorial(3) = 3 * factorial(2)

File: Recursion_and_Backtracking_pepcoding_level_1/1.Recursion_Intro/4.factorial.rb
# def factorial n
#   return 1 if n == 1

#   return n * factorial(n-1)
# end

def factorial(n) n == 1 ? 1 : n * factorial(n - 1) end


puts factorial 5

# // Expectation
# // factorial(3) = 3 * 2 * 1;

# // Faith
# // factorial(2) = 2 * 1;

# // F + E
# // factorial(3) = 3 * factorial(2)
